### DISCUSSION TITLE
PHP Extensions

### DISCUSSION THREAD

**Discussion by AlexOstrovsky** [pkgxdev/pkgx#839] [Help]
Is it possible to use pkgx to install php and extensions e.g. rdkafka, redis, apcu?

**Comment by jhheider** [comment]
We certainly have those libraries. I haven't dug into php extensions in many years, but I suspect they need to be built as a piece of intermediary code. If so, it's likely its own package, which could be packaged.

**Reply by AlexOstrovsky** [reply]
currently on homebrew I am using either this: https://github.com/shivammathur/homebrew-extensions or istalling via https://pecl.php.net

**Reply by AlexOstrovsky** [reply]
@jhheider is there any way how to do this? I would like to use pkgx on localhost as dev environment and also in github pipeline to run builds, but without php extensions I cannot.

**Reply by jhheider** [reply]
The right way is probably to package PEAR, yes? I can look into it.

**Reply by AlexOstrovsky** [reply]
@jhheider I think you mean PECL and yeah that would be great if you could look into it.

**Reply by jhheider** [reply]
yes, i think that's correct. it's been almost a decade, so my php terminology is rusty. i was happy when composer got popular for making things a little more modern and a little less compile-from-c.

**Reply by jhheider** [reply]
built pear/pecl for 8.2.12: https://github.com/pkgxdev/pantry/actions/runs/6715347278

I'll trigger rebuilds for the other versions we have; they take about 3 hours each. I'll try and get ~8.1, ~8.0, and ~7.4 done by morning. I'll rebuild the rest over time.

You can replace the version in place by removing it from your ~/.pkgx directory and getting it again.

**Reply by AlexOstrovsky** [reply]
@jhheider what does it mean? 😃

**Reply by jhheider** [reply]
The newest 8.0,8.1, and 8.2 builds have pecl/pear built now. Others will need to be rebuilt (but they're slow)

**Reply by AlexOstrovsky** [reply]
and how do I use it in e.g. pkgx.yaml file?

**Reply by jhheider** [reply]
any environment that has `php.net` in it, should have `pecl` and `pear`. you should be able to call them as usual.

**Reply by AlexOstrovsky** [reply]
Can you please advise what I am doing wrong: I have pkgx.yaml file with dependency php@8.2.12. But when I run `pecl -v` I get: `/Users/alex.ostrovsky/.pkgx/php.net/v8.2.12/bin/pecl: line 28: /opt/php.net/v8.2.12/bin/php: No such file or directory
/Users/alex.ostrovsky/.pkgx/php.net/v8.2.12/bin/pecl: line 28: exec: /opt/php.net/v8.2.12/bin/php: cannot execute: No such file or directory` .

`which pecl`:
`/Users/alex.ostrovsky/.pkgx/php.net/v8.2.12/bin/pecl`

`which php`:
`/Users/alex.ostrovsky/.pkgx/php.net/v8.2.12/bin/php`

Also one more question, would it be possible for pkgx to parse composer.json file and automatically install extensions defined there and also take php version defined there? Or should I open this in another discussion?

**Reply by jhheider** [reply]
looks like i missed a bunch of embedded strings. i'll push a fix to that today. `composer.json` support in `pkgx` should certainly be possible. opening an issue on `pkgxdev/pkgx` will help with visibility, spec generation, and marshaling labor if needed.

if you want to try the fix sooner, you can do this:
```shell
sed -i.bak -e's|/opt/php.net/v8.2.12|$(dirname "$(dirname "$0")")|g' ~/.pkgx/php.net/v8.2.12/bin/{pear,peardev,pecl}
```

**Reply by AlexOstrovsky** [reply]
@jhheider when I try to install extension with pecl I get: 
```
pecl install amqp
Cannot install, php_dir for channel "pecl.php.net" is not writeable by the current user
```

when I try with sudo I get: 
```
sudo pecl install amqp
No releases available for package "pecl.php.net/amqp"
install failed
```

Can you please advise?
